drivers/flash/soc_flash_nrf_mram.c

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+
+#include <zephyr/drivers/flash.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/barrier.h>
+
+LOG_MODULE_REGISTER(flash_nrf_mram, CONFIG_FLASH_LOG_LEVEL);
+
+#define DT_DRV_COMPAT nordic_mram
+
+#define MRAM_START DT_INST_REG_ADDR(0)
+#define MRAM_SIZE  DT_INST_REG_SIZE(0)
+
+#define MRAM_WORD_SIZE DT_INST_PROP(0, write_block_size)
+#define MRAM_WORD_MASK (MRAM_WORD_SIZE - 1)
+
+#define ERASE_VALUE 0xff
+
+#if defined(CONFIG_SOC_FLASH_NRF_MRAM_ONE_BYTE_WRITE_ACCESS)
+#define WRITE_BLOCK_SIZE 1
+#else
+#define WRITE_BLOCK_SIZE MRAM_WORD_SIZE
+#endif
+
+#define PAGE_SIZE  CONFIG_SOC_FLASH_NRF_MRAM_LAYOUT_PAGE_SIZE
+#define PAGE_COUNT ((MRAM_SIZE) / (PAGE_SIZE))
+
+BUILD_ASSERT(MRAM_START > 0, "nordic,mram: start address expected to be non-zero");
+BUILD_ASSERT(IS_POWER_OF_TWO(MRAM_WORD_SIZE),
+	     "nordic,mram: write-block-size expected to be a power of 2");
+BUILD_ASSERT((PAGE_SIZE % WRITE_BLOCK_SIZE) == 0,
+	     "page size expected to be a multiple of write block size");
+
+/**
+ * @param[in,out] offset      Relative offset into memory, from the driver API.
+ * @param[in]     len         Number of bytes for the intended operation.
+ * @param[in]     must_align  Require MRAM word alignment, if applicable.
+ *
+ * @return Absolute address in MRAM, or NULL if @p offset or @p len are not
+ *         within bounds or appropriately aligned.
+ */
+static uintptr_t validate_and_map_addr(off_t offset, size_t len, bool must_align)
+{
+	if (unlikely(offset < 0 || offset + len > MRAM_SIZE)) {
```

Comment:
I think that correct would be to do `offset < 0 || offset >= MRAM_SIZE || ((MRAM_SIZE - offset) < len)`, as this check `offset` against two constant boundaries we know, and then `len` against already verified range.
This also guards against `offset + len` overflowing to something that will be `< MRAM_SIZE`.

Reply:
Updated, thanks